Add clarification for partial read breaking change (#40057)
dotnet/docs: docs/core/compatibility/core-libraries/6.0/partial-byte-reads-in-streams.md

---
title: "Breaking change: Partial and zero-byte reads in DeflateStream, GZipStream, and CryptoStream"
description: Learn about the .NET 6 breaking change in core .NET libraries where DeflateStream, GZipStream, and CryptoStream handle partial and zero-byte reads differently.
ms.date: 03/18/2024
---
# Partial and zero-byte reads in DeflateStream, GZipStream, and CryptoStream

> The `Read()` and `ReadAsync()` methods on <xref:System.IO.Compression.DeflateStream>, <xref:System.IO.Compression.GZipStream>, and <xref:System.Security.Cryptography.CryptoStream> might no longer return all the requested bytes.

Reviewer: ```suggestion
The `Read()` and `ReadAsync()` methods on <xref:System.IO.Compression.DeflateStream>, <xref:System.IO.Compression.GZipStream>, and <xref:System.Security.Cryptography.CryptoStream> may no longer return as many bytes as were requested.
```

---
title: "Breaking change: Partial and zero-byte reads in DeflateStream, GZipStream, and CryptoStream"
description: Learn about the .NET 6 breaking change in core .NET libraries where DeflateStream, GZipStream, and CryptoStream handle partial and zero-byte reads differently.
ms.date: 03/18/2024
---
# Partial and zero-byte reads in DeflateStream, GZipStream, and CryptoStream

> The `Read()` and `ReadAsync()` methods on <xref:System.IO.Compression.DeflateStream>, <xref:System.IO.Compression.GZipStream>, and <xref:System.Security.Cryptography.CryptoStream> might no longer return as many bytes as were requested.

Reviewer 2: `might no longer` sounds odd to me. What about
```suggestion
The `Read()` and `ReadAsync()` methods on <xref:System.IO.Compression.DeflateStream>, <xref:System.IO.Compression.GZipStream>, and <xref:System.Security.Cryptography.CryptoStream> might not return as many bytes as were requested.
```

Author: With "might not", it's unclear if it's referring to the old behavior or the new behavior.

```output
Read: 150 bytes
Total received: 150 bytes
```

In .NET 6 and later versions, the following output shows that `Read` was called multiple times until *all* the requested data was received. Even though the call to `Read` requests 150 bytes, each call to `Read` was able to successfully decompress *some* bytes (that is, all of the bytes that had been received at that time) to return, and it did:

```output
Read: 1 bytes
Read: 101 bytes
Read: 4 bytes
Read: 4 bytes
Read: 2 bytes
Read: 2 bytes
Read: 2 bytes
Read: 2 bytes
Read: 3 bytes
Read: 2 bytes
Read: 3 bytes
Read: 2 bytes
Read: 2 bytes
Read: 2 bytes
Read: 2 bytes
Read: 1 bytes
Read: 2 bytes
Read: 1 bytes
Read: 1 bytes
Read: 1 bytes
Read: 2 bytes
Read: 1 bytes
Read: 1 bytes
Read: 2 bytes
Read: 1 bytes
Read: 1 bytes
Read: 2 bytes
Total received: 150 bytes
```

## Old behavior

When `Stream.Read` or `Stream.ReadAsync` was called on one of the affected stream types with a buffer of length `N`, the operation wouldn't complete until:

- `N` bytes had been read from the stream, or

> - The underlying stream they wrap returned 0 from a call to its read, indicating no more data was available.

Reviewer 2: I think the phrase `underlying stream` indicates it's wrapped, so this is redundant?
```suggestion
- The underlying stream returned 0 from a call to its read, indicating no more data was available.
```

Also, when `Stream.Read` or `Stream.ReadAsync` was called with a buffer of length 0, the operation would succeed immediately, sometimes without doing a zero-length read on the stream it wraps.

## New behavior

Starting in .NET 6, when `Stream.Read` or `Stream.ReadAsync` is called on one of the affected stream types with a buffer of length `N`, the operation completes when:

- At least 1 byte has been read from the stream, or

> - The underlying stream they wrap returns 0 from a call to its read, indicating no more data is available.

Reviewer 2: This one doesn't have the same changes as made in the **old behavior** section.

Author: Do you mean there's no behavior difference between old/new?

Starting in .NET 6, when `Stream.Read` or `Stream.ReadAsync` is called on one of the affected stream types with a buffer of length `N`, the operation completes when:

- At least 1 byte has been read from the stream, or
- The underlying stream they wrap returns 0 from a call to its read, indicating no more data is available.

Also, when `Stream.Read` or `Stream.ReadAsync` is called with a buffer of length 0, the operation succeeds once a call with a nonzero buffer would succeed.

> For example, the following call to <xref:System.IO.Compression.GZipStream.Read%2A?displayProperty=nameWithType> doesn't read all of the compressed text for very long strings.

Author: @stephentoub I'm making this update in response to a request from a top/paid customer. Is there a numerical value I can use here instead of "very long"?

Reviewer: This isn't about "very long strings". It's purely about whether there's data to be read. When you call Read, if the Read can satisfy at least one byte of the request, regardless of how many were requested, it will return as many as it can at that moment. Consider this example:
```C#
using System.IO;
using System;
using System.IO.Compression;
using System.Net;
using System.Net.Sockets;
using System.Threading.Tasks;

internal class Program
{
    private static async Task Main()
    {
        // Connect two sockets and wrap a stream around each
        using (Socket listener = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
        using (Socket client = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
        {
            listener.Bind(new IPEndPoint(IPAddress.Loopback, 0));
            listener.Listen(int.MaxValue);
            client.Connect(listener.LocalEndPoint);
            using (Socket server = listener.Accept())
            {
                var clientStream = new NetworkStream(client, ownsSocket: true);
                var serverStream = new NetworkStream(server, ownsSocket: true);

                // Create some compressed data
                var compressedData = new MemoryStream();
                using (var gz = new GZipStream(compressedData, CompressionLevel.Fastest, leaveOpen: true))
                {
                    byte[] bytes = new byte[1000];
                    new Random().NextBytes(bytes);
                    gz.Write(bytes, 0, bytes.Length);
                }

                // Trickle it from the client stream to the server
                Task sendTask = Task.Run(() =>
                {
                    foreach (byte b in compressedData.ToArray())
                    {
                        clientStream.WriteByte(b);
                    }
                    clientStream.Dispose();
                });

                // Read and decompress all the sent bytes
                byte[] buffer = new byte[1_000];
                int total = 0;
                using (var gz = new GZipStream(serverStream, CompressionMode.Decompress))
                {
                    int numRead;
                    while ((numRead = gz.Read(buffer, 0, buffer.Length)) > 0)
                    {
                        total += numRead;
                        Console.WriteLine($"Read: {numRead}");
                    }
                }
                Console.WriteLine("Total received: " + total);

                await sendTask;
            }
        }
    }
}
```
That creates and compresses a thousand random bytes. It then sends the compressed data one byte at a time from the client to the server, with the server decompressing the data with calls to Read 1000 bytes. On .NET Framework (and older .NET Core versions), that would produce this output:
```
Read: 1000
Total received: 1000
```
because even though data is available for GZipStream to return, it's forced to wait until the requested number of bytes is available.  With .NET 6+, you instead get output along the lines of this:
```
Read: 1
Read: 110
Read: 3
Read: 4
Read: 4
Read: 4
Read: 3
Read: 4
Read: 3
Read: 4
Read: 3
Read: 3
Read: 5
Read: 4
Read: 4
Read: 3
Read: 1
Read: 3
Read: 3
Read: 3
Read: 3
Read: 2
Read: 3
Read: 3
Read: 3
Read: 3
Read: 3
Read: 2
Read: 3
Read: 4
Read: 6
Read: 3
Read: 5
Read: 4
Read: 5
Read: 4
Read: 5
Read: 5
Read: 5
Read: 5
Read: 5
Read: 3
Read: 6
Read: 4
Read: 5
Read: 3
Read: 7
Read: 6
Read: 5
Read: 6
Read: 4
Read: 4
Read: 2
Read: 5
Read: 4
Read: 4
Read: 4
Read: 4
Read: 4
Read: 3
Read: 5
Read: 4
Read: 3
Read: 1
Read: 1
Read: 4
Read: 14
Read: 5
Read: 2
Read: 2
Read: 3
Read: 4
Read: 5
Read: 4
Read: 6
Read: 6
Read: 6
Read: 6
Read: 6
Read: 9
Read: 9
Read: 8
Read: 10
Read: 10
Read: 7
Read: 5
Read: 10
Read: 6
Read: 6
Read: 6
Read: 6
Read: 7
Read: 6
Read: 6
Read: 6
Read: 6
Read: 6
Read: 8
Read: 8
Read: 8
Read: 8
Read: 4
Read: 6
Read: 6
Read: 3
Read: 3
Read: 5
Read: 5
Read: 10
Read: 6
Read: 6
Read: 5
Read: 3
Read: 6
Read: 6
Read: 5
Read: 5
Read: 6
Read: 6
Read: 8
Read: 7
Read: 6
Read: 6
Read: 10
Read: 5
Read: 6
Read: 5
Read: 6
Read: 5
Read: 5
Read: 5
Read: 5
Read: 5
Read: 7
Read: 6
Read: 5
Read: 5
Read: 5
Read: 4
Read: 6
Read: 5
Read: 6
Read: 5
Read: 5
Read: 5
Read: 5
Read: 5
Read: 5
Read: 5
Read: 6
Read: 5
Read: 5
Read: 4
Read: 5
Read: 6
Read: 5
Read: 5
Read: 5
Read: 6
Read: 4
Read: 6
Read: 5
Read: 6
Read: 5
Read: 5
Read: 6
Read: 4
Read: 5
Read: 5
Read: 5
Read: 5
Read: 6
Read: 5
Read: 5
Read: 5
Read: 5
Read: 6
Read: 5
Read: 2
Total received: 1000
```
because even though the call to Read is requesting 1000 bytes, the Read call was able to successfully decompress received bytes into decompressed bytes that can be returned, and it does.

Also, in highly scalable applications, it's common to use zero-byte reads as a way of delaying buffer allocation until a buffer is needed. An application can issue a read with an empty buffer, and when that read completes, data should soon be available to consume. The application can then issue the read again, this time with a buffer to receive the data. By delegating to the wrapped stream if no already decompressed or transformed data is available, these streams now inherit any such behavior of the streams they wrap.

## Recommended action

In general, code should:

- Not make any assumptions about a stream `Read` or `ReadAsync` operation reading as much as was requested. The call returns the number of bytes read, which might be less than what was requested. If an application depends on the buffer being completely filled before progressing, it can perform the read in a loop to regain the behavior.

  ```csharp
  int totalRead = 0;
  while (totalRead < buffer.Length)
  {

>       int bytesRead = stream.Read(buffer, totalRead, buffer.Length - totalRead);

Reviewer: Why are we changing this?

>       int bytesRead = stream.Read(buffer, totalRead, buffer.Length - totalRead);

Author: I got a compile error when I tried to use Slice().

Reviewer: Ah, it should be `AsSpan(totalRead)`, not `Slice(totalRead)`.

Author: Thanks, that worked.
